### PR TITLE
[#1503] feat(IT): Gravitino web e2e test framework

### PR DESCRIPTION
**Pull request by xunliu**
### What changes were proposed in this pull request?

1. Automatic download `Chrome` & `Chrome Driver` into `./intergration-test/build/chrome` directory. need to download about 150 MB zip file for the first time.
2. Support embedded and deploy test mode.
3. Support Linux and macOS operation systems
4. Successed load metalake webpage and check the browser title.
5. #1687 

### Why are the changes needed?

Gravitino Web UI needs an E2E test to ensure Web UI usability.

Fix: #1503

### Does this PR introduce _any_ user-facing change?

N/A

### How was this patch tested?

CI Passed

### CODE REVIEW

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/ChromeWebDriverProvider.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    } else if (SystemUtils.IS_OS_WINDOWS) {
+      this.chromeDriverBinName = ITUtils.joinPath("chromedriver_win32", "chromedriver.exe");
+      this.chromeBinName = ITUtils.joinPath("chrome-win", "chrome.exe");
+    } else {
+      throw new RuntimeException("Unsupported OS");
+    }
+  }
+
+  @Override
+  public void downloadWebDriver() {
```
Comment (diqiu50) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/ChromeWebDriverProvider.java:45]: Why aren't these downloads implemented with a separate script?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/WebDriverManager.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      } catch (TimeoutException e) {
+        LOG.info("Exception in WebDriverManager while WebDriverWait ", e);
+        driver.navigate().to(url);
+      }
+    }
+
+    if (!loaded) {
+      fail();
+    }
+    Dimension d = new Dimension(1920, 1080);
```
Comment (diqiu50) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/WebDriverManager.java:56]: The resolution here should not be hardcoded, and 1920*1080 is too large; many monitors are not that big. Let's ask @ch3yne  to determine an appropriate size.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/WebDriverManager.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        loaded = true;
+        break;
+      } catch (TimeoutException e) {
+        LOG.info("Exception in WebDriverManager while WebDriverWait ", e);
+        driver.navigate().to(url);
+      }
+    }
+
+    if (!loaded) {
+      fail();
```
Comment (diqiu50) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/WebDriverManager.java:54]: Please give a clear error message.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/WebDriverManager.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    boolean loaded = false;
+    while (System.currentTimeMillis() - start < 60 * 1000) {
+      try {
+        (new WebDriverWait(driver, 10))
+            .until(
+                new ExpectedCondition<Boolean>() {
+                  @Override
+                  public Boolean apply(WebDriver d) {
+                    String gravitinoVersion = d.findElement(By.id("gravitino_version")).getText();
+                    String projectVersion = System.getenv("PROJECT_VERSION");
```
Comment (diqiu50) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/WebDriverManager.java:41]: `PROJECT_VERSION` should not retrieved by gravitino server api

**Review by diqiu50 — COMMENTED** [review]
I suggest running web integration tests through special command arguments.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/WebDriverManager.java (hunk 3 above)
Comment (yuqi1129) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/WebDriverManager.java:54]: +1, Please do not use Junit 4 API

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/ChromeWebDriverProvider.java (hunk 1 above)
Comment (xunliu) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/ChromeWebDriverProvider.java:45]: I think better use JAVA method to download web browser driver.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/WebDriverManager.java (hunk 2 above)
Comment (ch3yne) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/WebDriverManager.java:56]: `1440*1080` or `1440*900` is usually enough.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/WebDriverManager.java (hunk 4 above)
Comment (xunliu) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/WebDriverManager.java:41]: But I worry Gravitino server failed to start.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/WebDriverManager.java (hunk 2 above)
Comment (xunliu) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/WebDriverManager.java:56]: @ch3yne Thank you.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/WebDriverManager.java (hunk 3 above)
Comment (xunliu) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/WebDriverManager.java:54]: I accepted.

File: integration-test/build.gradle.kts [hunk 5]
```diff
@@ -269,6 +271,7 @@ tasks.test {
 
       // Default use MiniGravitino to run integration tests
       environment("GRAVITINO_ROOT_DIR", rootDir.path)
+      environment("IT_PROJECT_DIR", buildDir.path)
```
[the remark this answers is not in the document]
Comment (xunliu) [reply integration-test/build.gradle.kts:274]: Avoid splice integration-test module path, and hard code `System.getenv("GRAVITINO_ROOT_DIR") + '\' + 'integration-test'`

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/AbstractWebIT.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+package com.datastrato.gravitino.integration.test.web.ui.utils;
+
+import com.datastrato.gravitino.integration.test.util.AbstractIT;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.openqa.selenium.WebDriver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AbstractWebIT extends AbstractIT {
```
[the remark this answers is not in the document]
Comment (xunliu) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/AbstractWebIT.java:14]: DONE

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/ChromeWebDriverProvider.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+      this.chromeDriverBinName = ITUtils.joinPath("chromedriver_mac64", "chromedriver");
+      this.chromeBinName = ITUtils.joinPath("chrome-mac", "Chromium.app");
+    } else if (SystemUtils.IS_OS_LINUX) {
+      this.chromeDriverBinName = ITUtils.joinPath("chromedriver_linux64", "chromedriver");
+      this.chromeBinName = ITUtils.joinPath("chrome-linux", "chrome");
+    } else if (SystemUtils.IS_OS_WINDOWS) {
+      this.chromeDriverBinName = ITUtils.joinPath("chromedriver_win32", "chromedriver.exe");
+      this.chromeBinName = ITUtils.joinPath("chrome-win", "chrome.exe");
+    } else {
+      throw new RuntimeException("Unsupported OS");
```
[the remark this answers is not in the document]
Comment (xunliu) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/ChromeWebDriverProvider.java:40]: Good idea.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/ChromeWebDriverProvider.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    StringBuilder downloadURL =
+        new StringBuilder(
+            "https://www.googleapis.com/download/storage/v1/b/chromium-browser-snapshots/o/");
+    StringBuilder chromeDownloadURL = new StringBuilder(downloadURL),
+        chromeDriverDownloadURL = new StringBuilder(downloadURL);
+
+    String chromeZipFile = "", chromeDriverZipFile = "";
+    if (SystemUtils.IS_OS_LINUX) {
+      // https://www.googleapis.com/download/storage/v1/b/chromium-browser-snapshots/o/Linux_x64%2F1000022%2Fchrome-linux.zip?generation=1651778257041732&alt=media
+      chromeZipFile = "chrome-linux.zip";
```
[the remark this answers is not in the document]
Comment (xunliu) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/ChromeWebDriverProvider.java:56]: I accepted your suggestion.

**Comment by xunliu** [comment]
@yuqi1129 Please help me review this PR again, Thanks.

**Review by yuqi1129 — APPROVED** [review]
LGTM

**Review by yuqi1129 — APPROVED** [review]
LGTM

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/WebDriverManager.java (hunk 4 above)
Comment (diqiu50) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/ui/utils/WebDriverManager.java:40]: If the Gravitino server failed to start. The tester also failed. So it does't matter

**Pull request merged**